### ISSUE TITLE
TypeError - object of type 'UnrecognizedExtension' has no len()

### ISSUE THREAD

**Issue by blshkv** [nabla-c0d3/sslyze#320]
python2 sslyze --regular www.pentest.co.uk

```
 * Certificate Information:
Unhandled exception while running --certinfo:
TypeError - object of type 'UnrecognizedExtension' has no len()
```
net-analyzer/sslyze-1.4.2
dev-python/nassl-1.1.3

**Comment by nabla-c0d3** [comment]
Which version of `pyca/cryptography` has been installed?

**Comment by blshkv** [comment]
```
dev-python/cryptography-2.2.2
```

**Comment by blshkv** [comment]
ping?

**Comment by Glitched** [comment]
I'm also having this issue with the same versions

**Comment by blshkv** [comment]
I tried to downgrade to sslyze-1.3.4, nassl-1.0.3, cryptography-2.1.4. The error is still there.

**Comment by nabla-c0d3** [comment]
It looks like cryptography does not always support parsing Certificate Transparency extensions, perhaps depending on the OpenSSL version that's available. I've asked for more details on the cryptography IRC channel.

**Comment by blshkv** [comment]
the commit fa32ee143ef8e04ac86e97c2f698e19e062a2c97 changed nothing. I'm still getting the same error

**Comment by blshkv** [comment]
yes, the error is exactly the same, I suspect a different (more recent?) dependency, but I can't figure out which one.

```
bash$ python3.6 -m sslyze --certinfo www.pentest.co.uk



 AVAILABLE PLUGINS
 -----------------

  CertificateInfoPlugin
  OpenSslCcsInjectionPlugin
  RobotPlugin
  FallbackScsvPlugin
  HttpHeadersPlugin
  SessionResumptionPlugin
  SessionRenegotiationPlugin
  OpenSslCipherSuitesPlugin
  HeartbleedPlugin
  CompressionPlugin



 CHECKING HOST(S) AVAILABILITY
 -----------------------------

   www.pentest.co.uk:443                       => 31.193.5.173 




 SCAN RESULTS FOR WWW.PENTEST.CO.UK:443 - 31.193.5.173
 -----------------------------------------------------

 * Certificate Information:
Unhandled exception while running --certinfo:
TypeError - object of type 'UnrecognizedExtension' has no len()


 SCAN COMPLETED IN 2.46 S
 ------------------------

```

**Comment by blshkv** [comment]
I tried to upgrade few python packages to the latest, listed in https://github.com/nabla-c0d3/sslyze/blob/drop-python2-support/Pipfile.lock
It did not help either..

**Comment by blshkv** [comment]
ok, found it!
https://github.com/nabla-c0d3/sslyze/commit/fa32ee143ef8e04ac86e97c2f698e19e062a2c97#diff-91184cf30395d747bc33b5eea5149c57R121
```
-            if isinstance(sct_ext, cryptography.x509.UnrecognizedExtension):
+           if isinstance(sct_ext.value, cryptography.x509.UnrecognizedExtension):
```

Can you explain which version of openssl is too old?
I believe sslyze uses its own build-in versions, no?

openssl on both my system and a remove server is up-to-dated.

**Comment by nabla-c0d3** [comment]
Thanks for looking into this; I just pushed your fix. Can you confirm that it works?
As for OpenSSL, yes sslyze (more specifically nassl) embeds two versions of OpenSSL. However the code in this issue uses the cryptoraphy module to parse certificates, and this module uses the system's OpenSSL.

**Comment by blshkv** [comment]
ok, confirm the issue is fixed.

as for the openssl, I have dev-libs/openssl-1.0.2o on both client and server, so the message about "old" openssl is confusing. sslyze should not hit that exception since this version of openssl should support the extension.

**Comment by nabla-c0d3** [comment]
Cool to know that it's fixed. Thanks for help!

For the OpenSSL version, I don't know. The cryptography module is unable to parse the Certificate Transparency extension which is why it returns an UnrecognizedExtension; this is a problem in cryptography and not SSLyze. I asked on their IRC channel and they said that if the OpenSSL that was used with cryptography does not support CT, cryptography would not able to parse the CT extension.

**Comment by blshkv** [comment]
Ok, thanks. I'll try to investigate further

**Comment by nabla-c0d3** [comment]
Released as part of [2.0.0](https://github.com/nabla-c0d3/sslyze/releases/tag/2.0.0).
